### PR TITLE
Feature: Essence Shop & Carnival Shop Helper

### PR DESCRIPTION
**Pull request by DavidArthurCole**
## Dependencies
Essence shop data already existed in NEU repo, but carnival token costs were new so:
- https://github.com/NotEnoughUpdates/NotEnoughUpdates-REPO/pull/1375

## What
Adds a helper to Essence Shops and Carnival Token Shops that tells you how much essence/how many tokens you need to max all upgrades in the shop.

<details>
<summary>Images</summary>

Essence shops:

image

image

image

image

Carnival token shops:

image

image

</details>

## Changelog New Features
+ Added helpers for Essence Shops and Carnival Event Upgrade Shops. - Daveed
    * Assists with maxing upgrades.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+            event.replace(progressStack)
+        }
+    }
+
+    @SubscribeEvent
+    fun onSlotClick(event: GuiContainerEvent.SlotClickEvent) {
+        if (currentProgress == null || event.slotId != CUSTOM_STACK_LOCATION || currentEssenceItem == null) return
+        event.cancel()
+        if (lastClick.passedSince() > 0.3.seconds) {
+            BazaarApi.searchForBazaarItem(currentEssenceItem!!, essenceNeeded)
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt]: you can do `val currentEssenceItem = currentEssenceItem  ?: return` here so you dont need to do the double bang

File: src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        }
+    }
+
+    private fun processEssenceShopHeader(event: InventoryOpenEvent) {
+        val essenceHeaderStack = event.inventoryItems[4]
+        if (essenceHeaderStack == null || !essenceShopPattern.matches(essenceHeaderStack.displayName))
+            ErrorManager.skyHanniError("Could not read current Essence Count from inventory §c§l${event.inventoryName}")
+
+        currentEssenceCountPattern.firstMatcher(essenceHeaderStack.getLore()) {
+            ownedEssence = groupOrNull("count")?.replace(",", "")?.toInt() ?: 0
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt]:
```suggestion
            ownedEssence = groupOrNull("count")?.formatInt() ?: 0
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+         * Filter out items that fall outside the bounds of 10 - 33
+         */
+        val upgradeStacks = inventoryItems.filter { it.key in 10..33 && it.value.item != null }
+        val purchasedUpgrades: MutableMap<String, Int> = buildMap {
+            upgradeStacks.forEach {
+                essenceUpgradePattern.matchMatcher(it.value.displayName) {
+                    val upgradeName = groupOrNull("upgrade") ?: return
+                    val nextUpgradeRoman = groupOrNull("tier") ?: return
+                    val nextUpgrade = nextUpgradeRoman.romanToDecimal()
+                    val isMaxed = it.value.getLore().any { loreLine -> Regex(".*§a§lUNLOCKED").containsMatchIn(loreLine) }
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt]: this should probably also be a repopattern

File: src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            essenceShops.find { it.shopName == essenceName } ?: return
+            processEssenceShopUpgrades(essenceName, event.inventoryItems)
+            processEssenceShopHeader(event)
+        }
+    }
+
+    private fun processEssenceShopHeader(event: InventoryOpenEvent) {
+        val essenceHeaderStack = event.inventoryItems[4]
+        if (essenceHeaderStack == null || !essenceShopPattern.matches(essenceHeaderStack.displayName))
+            ErrorManager.skyHanniError("Could not read current Essence Count from inventory §c§l${event.inventoryName}")
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt]: A `ErrorManager.logErrorWithData()` with fitting extra data is probably better

**Review by j10a1n15 — CHANGES_REQUESTED** [review]

**Comment by DavidArthurCole** [comment]
NEU data is fixed, all shops tested and looking good

image

File: src/main/java/at/hannibal2/skyhanni/data/jsonobjects/repo/EssenceShopJson.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/data/jsonobjects/repo/EssenceShopJson.kt:1]: For consistancy his file should be in the neu folder of json objects and the class name should start with Neu

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

**Comment by DavidArthurCole** [comment]
Adding support for Carnival Shops in this PR as well, since the code shards are almost nearly identical - apart from the fact that carnival tokens can't be bought/sold, making this one significantly easier.
image

**Review by hannibal002 — CHANGES_REQUESTED** [review]
works perfectly fine!
two small things:
please remove the italic effect on the yellow command line. this is not matching hypixel or skyhannis design pattern.
please add the gray line "fron skyhanni" just like for every other custom menu item.
image
image

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**